transformations.go

```diff
@@ -0,0 +1,142 @@
+package main
+
+import (
+	"math/rand"
+	"time"
+
+	weightedRand "github.com/mroth/weightedrand"
+)
+
+// Codon holds information for a codon triplet in a struct
+type Codon struct {
+	Triplet    string
+	Occurrence int
+}
+
+// AminoAcid holds information for an amino acid and related codons in a struct in a struct
+type AminoAcid struct {
+	Letter string
+	Codons []Codon
+}
+
+// CodonTable holds information for a codon table.
+type CodonTable struct {
+	StartCodons []string
+	StopCodons  []string
+	AminoAcids  []AminoAcid
+}
+
+// Generate map of amino acid -> codon chooser
+func (codonTable CodonTable) generateOptimizationTable() map[string]weightedRand.Chooser {
+	rand.Seed(time.Now().UTC().UnixNano())
+	var optimizationMap = make(map[string]weightedRand.Chooser)
+	for _, aminoAcid := range codonTable.AminoAcids {
+		// Get list of triplets and their weights
+		codonChoices := make([]weightedRand.Choice, len(aminoAcid.Codons))
+		totals := 0
+		for _, codon := range aminoAcid.Codons {
+			codonChoices = append(codonChoices, weightedRand.Choice{Item: codon.Triplet, Weight: uint(codon.Occurrence)})
+			totals += codon.Occurrence
+		}
+		optimizationMap[aminoAcid.Letter] = weightedRand.NewChooser(codonChoices...)
+	}
+	return optimizationMap
+}
+
+// Optimize takes an amino acid sequence and CodonTable and returns an optimized codon sequence
+func Optimize(aminoAcids string, codonTable CodonTable) string {
+	var codons string
+	optimizationTable := codonTable.generateOptimizationTable()
+	for _, aminoAcid := range aminoAcids {
+		codons += optimizationTable[string(aminoAcid)].Pick().(string)
+	}
+	return codons
+}
+
+// Generate map of codons -> amino acid
+func (codonTable CodonTable) generateTranslationTable() map[string]string {
+	var translationMap = make(map[string]string)
+	for _, aminoAcid := range codonTable.AminoAcids {
+		for _, codon := range aminoAcid.Codons {
+			translationMap[codon.Triplet] = aminoAcid.Letter
+		}
+	}
+	return translationMap
+}
+
+// Translate translates a codon sequence to an amino acid sequence
+func Translate(nucleotides string, codonTable CodonTable) string {
+	var aminoAcids string
+	translationTable := codonTable.generateTranslationTable()
+	length := len(nucleotides) / 3 // Assumes input sequences are divisible by 3
+	for i := 0; length > i; i++ {
+		aminoAcids += translationTable[nucleotides[i*3:(i+1)*3]]
+	}
+	return aminoAcids
+}
+
+// Function to generate default codon tables from NCBI https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi
+func generateCodonTable(aas, starts string) CodonTable {
```

Comment:
What does aas stand for here? Amino acid string?

Reply:
It stands for amino acids. On the link, they just name it "AAs", so I just copied their nomenclature (also why starts is named starts)